### ISSUE TITLE
VexFlow 5 Discussion

### ISSUE THREAD

**Issue by ronyeh** [vexflow/vexflow#13]
Lets discuss the overall direction of VexFlow 5 here.

Linking to some relevant previous discussions:
- https://github.com/0xfe/vexflow/issues/1498
- https://github.com/0xfe/vexflow/discussions/1577

**Comment by ronyeh** [comment]
@rvilarl Thanks for submitting the first PR for VexFlow 5. I have begun reviewing it and will let you know when I'm done with my comments.

In the meantime, can you write a short paragraph or bullet points summarizing the main architectural changes you have worked on for VexFlow 5? Please highlight the main benefits (and notable drawbacks) of this new direction, so readers can have a chance to chime in with their feedback. 

Can you also link your example webapp that uses your branch of VexFlow 5, and show us how to see the new benefits of this approach. For example, we can go open developer tools and inspect how each component is loaded, and also inspect the SVG elements rendered onto the page.

Thanks for your help!

**Comment by ronyeh** [comment]
In the spirit of **"release early; release often"** I just published the first version of VexFlow 5 alpha to this location:

https://www.npmjs.com/package/vxflw-early-access

I didn't want to mess up our ability to patch the VexFlow 4.x LTS just yet... so we can keep publishing VexFlow 5 builds under this package name until we are happy with it.

Test with:
```
npm install vxflw-early-access
```

You might have to adjust your imports: 
```
import { Vex } from "vxflw-early-access";
```

**Comment by ronyeh** [comment]
@rvilarl Can you delete the `[vf5/vexflow5](https://github.com/vexflow/vexflow/tree/vf5/vexflow5)` branch from https://github.com/vexflow/vexflow/branches

I think we should keep the organization repository free from our personal branches if possible. (e.g., I also have a vexflow5 experimental branch on my own local working copy).

**Comment by rvilarl** [comment]
> @rvilarl Can you delete the `vf5/vexflow5` branch from https://github.com/vexflow/vexflow/branches
> 
> I think we should keep the organization repository free from our personal branches if possible. (e.g., I also have a vexflow5 experimental branch on my own local working copy).

@AaronDavidNewman this is the view on concept that you requested. You can still find it in my fork.

**Comment by ronyeh** [comment]
I just released a new alpha of v5.

You can now test the current main revision by doing:

```
npm install vxflw-early-access@5.0.0-alpha.1
```


The release steps below took ~4 mins of real time.  @rvilarl I invited you to maintain that early access npm package: https://www.npmjs.com/package/vxflw-early-access

Next time, I can walk you through this so you can make a release. (It's the same process once we are ready to publish the official v5 release to the VexFlow npm package).

```
git remote show origin
...
  Fetch URL: git@github.com:vexflow/vexflow.git
  Push  URL: git@github.com:vexflow/vexflow.git
```

```
git pull
```

```
git log
```

```
npm install
```

```
grunt release:alpha
```

... grunt builds the product ...

... In the meantime, I launch my 2FA app for npm publish ...

```
? Publish vxflw-early-access@alpha to npm? (Y/n) Y
? Please enter OTP for npm: xxxxx
? Commit (Release VexFlow 5.0.0-alpha.1)? Yes
? Tag (5.0.0-alpha.1)? Yes
? Push? Yes
? Create a pre-release on GitHub (5.0.0-alpha.1)? Yes
```

Output:
```
Successfully released vxflw-early-access 5.0.0-alpha.1 to https://github.com/vexflow/vexflow
🔗 https://www.npmjs.com/package/vxflw-early-access
🔗 https://github.com/vexflow/vexflow/releases/tag/5.0.0-alpha.1
```

**Comment by ronyeh** [comment]
@rvilarl Just checking in.... how far along are we in your conversion from v4 to v5? A rough % estimate is fine... so that others are aware. What are the big remaining things to convert before we start cleaning things up?

**Comment by rvilarl** [comment]
@ronyeh only one class before starting with the clean up!!!! 
I have two bugs in the conversion. Perhaps we can merge it with the bugs, document them as issues and we can start with cleaning things up.

**Comment by ronyeh** [comment]
I just released a new alpha of v5.

You can now test the current main revision by doing:

```
npm install vxflw-early-access@5.0.0-alpha.2
```

**Comment by ronyeh** [comment]
Next steps:

- We should work with Simon Tatham to get an official SMuFL build of Gonville integrated into VexFlow.
- I'll work on improving the API to `VexFlow.*` instead of `Vex.Flow.*`
- We should review any outstanding bugs https://github.com/vexflow/vexflow/issues/ and fix them if they are high enough priority.

After these three, we can promote V5 from Alpha ⇨ Beta.

Once in Beta we should promote it more and ask developers to try it out so we can find and fix more bugs so we can head toward the full release.

**Comment by ronyeh** [comment]
@jaredjj3 @mscuthbert @AaronDavidNewman @0xfe @sschmidTU Hello! :-)

Just tagging you all to see if you have any feature requests that you'd really really love to get into VexFlow 5.0. Last chance.... before @rvilarl and I start cleaning up and moving the repo to "feature freeze" so we can address bugs & minor details so we can eventually get this out the door.

Also, if we ever discover that we have removed something from V4 that is very important for your website or app, we will try our best to find a way to restore this feature. We'd like V5 to be awesome for everyone (including existing users of VexFlow).

**Comment by ronyeh** [comment]
The most recent alpha release was just published by Rodrigo. You can install it with:

```
 npm install vxflw-early-access@alpha
```

**Comment by 0xfe** [comment]
Hey, No particular feature requests, but very unclear about feature parity with the previous version.

Is there anything that we're giving up?

**Comment by rvilarl** [comment]
I think that we have kept all the features. 
We support now more fonts and all the glyphs of these fonts. V4 includes a subset of gyphs of a subset of these fonts.

**Comment by ronyeh** [comment]
> Is there anything that we're giving up?

Our goal was to keep features as much as possible, while 1) adding support for all SMuFL fonts and 2) modernizing the APIs to be more friendly to new developers (e.g., camelCase).

We have the same test case coverage as before, and have only modified them to work with the updated public API.

Here are some things that might be considered "lost" from V4:
- We no longer have glyph paths inside the library. We can no longer render these paths as curves. (Rodrigo removed much of the code that handles the drawing of glyphs.) The biggest drawback here is that if we use VexFlow as a node JS library to render out SVG XML files, these files will not be portable. Any user whose operating system that does not include the correct SMuFL fonts will see empty squares. We could fix this by having an optional peer dependency on opentype.js. If a developer wants to insert the glyph paths, they can call a VexFlow method that is available only if opentype.js is available at runtime. I know you were not a fan of external dependencies, but previous versions of VexFlow relied on opentype.js to extract the glyph paths. So now it would just be an optional runtime dependency, only if the developer really wants portable downloadable SVG images.
- The `vexflow.js` build target in V4 bundled Bravura, Petaluma, Gonville, Leland, and Custom. We have dropped Leland and Custom from the default build. We could also remove Gonville, but we kept it in since it's the classic VexFlow font.

If anyone finds that a feature they were using in V3 or V4 is no longer available in V5, Rodrigo and I will try our best to restore that functionality. Our goal with V5 was to improve the library while avoiding the loss of features as much as possible.

**Comment by rvilarl** [comment]
Hi Ron, I would not add opentype to make the svg portable. I would rather add to the svg the link to the font Url.

**Comment by rvilarl** [comment]
Is it a requirement to get vexfow to work offline? The fonts are also cached very well by the internet browsers. I will make a try to see what happens with this approach offline.

**Comment by ronyeh** [comment]
I had the impression that you wanted to be able to work offline. Wasn't there a situation earlier this year where you were on an airplane or something and the vexflow tests were not working?

In any case, an SVG with an embedded font link might work well. We would have to test if they work when you open the file in Adobe illustrator or other vector editors. But probably the "most portable" SVG files would be those with paths (like in V4). SVGs containing paths do not rely on a network connection or the presence of custom fonts.

**Comment by rvilarl** [comment]
> I had the impression that you wanted to be able to work offline. Wasn't there a situation earlier this year where you were on an airplane or something and the vexflow tests were not working?

yes I wanted to be able to code and test offline.
 
> In any case, an SVG with an embedded font link might work well. We would have to test if they work when you open the file in Adobe illustrator or other vector editors. But probably the "most portable" SVG files would be those with paths (like in V4). SVGs containing paths do not rely on a network connection or the presence of custom fonts.

I do not know but file size of svgs  with paths are not easy to handle. I would hope that an SVG with fonts should be portable. Let us further discuss in #71

**Comment by rvilarl** [comment]
@ronyeh I have published a prerelease. The change introduced in the last two merges is quite significant and I would like to avoid that new users start using vexflow without it. @jaredjj3 could you update vexml? vex.flow is now vexflow. See #136

**Comment by jaredjj3** [comment]
@rvilarl, thanks for the heads up. We should be good because we only use `import * as vexflow from 'vexflow'` throughout.

https://github.com/search?q=repo%3Astringsync%2Fvexml%20%22vex.flow%22&type=code

**Comment by rvilarl** [comment]
@ronyeh I believe that we could move to vexflow beta and leave the vexflow preview.

**Comment by ronyeh** [comment]
OK, let me integrate the fixed Gonville glyphs, and then we can use the `vexflow` project name instead of `vxflw-early-access`. That way more folks can test it out and help us find issues.

**Comment by ronyeh** [comment]
As Rodrigo said recently, Happy New Year to everyone in the VexFlow community! Next week I'll have time to finish the minor Gonville fixes and then push a prerelease to the npm 'vexflow' project. We can also work on the outstanding PRs so we can get the beta out to the wider community for testing.

**Comment by alindsay55661** [comment]
I am prepared to start deep testing smaller scores with interactivity if a release is available. In particular, is there documentation for the new apis in v5? I will be digging in fairly deep to address some things not native to the lib. Let me and thanks for all the fantastic work!

**Comment by ronyeh** [comment]
Thanks for helping out, Alan :-)

I'm fixing up a font right now, and hopefully I can ship a VexFlow alpha to NPM this week. Then we can use any & all testing (and development) help we can get from the community.

The goal for V5 was to be as compatible with V4 APIs as possible, while supporting a new way of loading SMuFL fonts. 

We definitely need help testing out the APIs and documenting them. But mostly the APIs are "documented" via the test cases in `vexflow/tests/*.ts`

**Comment by alindsay55661** [comment]
Ok great, will use the tests as my guide and look forward to the next release (I will be testing dynamic font switching as well).

**Comment by ronyeh** [comment]
I just shipped the current version as `vexflow@alpha`. See: https://www.npmjs.com/package/vexflow/v/5.0.0-alpha.2

```
npm install vexflow@alpha

```

FYI: I still need to update Gonville in this version, but Bravura and Petaluma should look fine.

Now we need to test test test test test, so we can promote it to beta / rc / full release.

**Comment by mscuthbert** [comment]
Congrats Ron! Huge appreciation!

One of my next direction requests is to focus on bringing back an important part of Vexflow 1 that got lost in 3 and 4 (partly because it wasn't tested) and that is to make sure that formatter never produces notes that extend beyond the end of the Stave even if the given Stave is too small to reasonably cover it. And algorithmally suggesting a proper Stave width given the ticks and tickcontexts. It's one of the most important parts of the Vexflow ecosystem that doesn't seem to be fully addressed.

**Comment by jaredjj3** [comment]
@mscuthbert, I'm very glad you brought this up — the timing couldn't have been better. I've been trying to fix the consequences of the issue you described in https://github.com/stringsync/vexml/issues/213, but I thought I was using `vexflow` incorrectly until I saw your comment.

When I use `vexflow` formatting to calculate a stave width, I don't have the issue. However, if a MusicXML document specifies a `measure[width]` smaller than the calculated width, I get the issue you described as the musical system width varies.

using `vxflw-early-access 5.0.0-alpha.11`

_system width 1296px_

<img width="1404" alt="image" src="https://github.com/vexflow/vexflow/assets/19232300/ca91068a-d0a3-46f5-bdd0-95bbbbac55b2">

_system width 1116px_

<img width="1153" alt="image" src="https://github.com/vexflow/vexflow/assets/19232300/b708a3d6-0b5a-4126-b69e-8033f64d6788">

Fortunately, I haven't sunk too much time trying to fix this within `vexml`. It sounds like I should wait for `vexflow 5` to support this better before chasing this down on my end.

**Comment by AaronDavidNewman** [comment]
> formatter never produces notes that extend beyond the end of the Stave even if the given Stave is too small to reasonably cover it

What is supposed to happen in this case?  I always thought having notes extend past the bar would be no worse than having them overlap.

I know the existing logic prioritizes notes appear in tick-order over fitting in the bar (e.g. the first g4 above will appear to the right of the e below it).  The logic will scooch the notes left-to-right to accomplish that, and then try to reformat into a smaller space if they go past the end.  It will also reformat them into a larger space if the right margin is too large.  And then give up if a certain number of iterations go by.

I'm glad you guys are taking a look at it.

**Comment by mscuthbert** [comment]
Thanks for taking the time to reply and think on this everyone!

> > formatter never produces notes that extend beyond the end of the Stave even if the given Stave is too small to reasonably cover it
> 
> What is supposed to happen in this case? I always thought having notes extend past the bar would be no worse than having them overlap.'

Here's why I think it's better to overlap than extend:

* If people are implementing the recommended method of making each measure its own stave, then there's still overlap: it just happens that the notes from one stave (measure) overlap the notes of the next measure.  This was the reason that first motivated the comment (the notes aren't being squished further by formatter because of the lyrics):

<img width="695" alt="Screenshot 2024-03-08 at 15 12 03" src="https://github.com/vexflow/vexflow/assets/3521479/e548388d-f0c7-4664-89cb-a47caacda786">

* A note beyond the end of the staff is always a mistake.  Overlapping tick contexts, however, is often not a mistake at all, especially if there is no graphical intersection.  For instance, the double-dot of a note on one staff might overlap with an accidental on the following note of the other staff (or its own staff at a different staff position).  A note w/ lyrics may overlap the below-staff lyrics of another voice.  These are not problems. 

* If notes are squished so much that there is graphical overlap (boo!), then at least reasonable X locations for all tick contexts have been computed and the programmer can use this information to adjust to alternative/non-preferred solutions for overlapping tickcontexts.  For instance, shifting rests up or down (would fix the Bach example above) moving accidentals closer to the notehead, substituting a narrower 8th note flag, writing two-notehead unisons w/ one notehead.  When the stave overflows, on the other hand, information about which graphics collide isn't available.
 
> I know the existing logic prioritizes notes appear in tick-order over fitting in the bar (e.g. the first g4 above will appear to the right of the e below it). The logic will scooch the notes left-to-right to accomplish that, and then try to reformat into a smaller space if they go past the end. It will also reformat them into a larger space if the right margin is too large. And then give up if a certain number of iterations go by.
> 
> I'm glad you guys are taking a look at it.

Thanks for the great comments.  I'll look into formatter later today.  One of the things I'm always trying to do is add more documentation as I look, so even if I can't solve the problem, maybe I can make it a bit better.

**Comment by judocode** [comment]
Hi all - I've been using Vexflow pretty extensively recently, love it, and just [fixed an issue with BarNotes not properly being represented as an Element](https://github.com/0xfe/vexflow/commit/c1d8ac265fb8d29a2930e4e36296a7635c2fe12c). I am excited about contributing more to this project as I further dig in. I figured I'd offer my perspective on what I think can be improved for v5 (and beyond) and I can definitely help contribute if folks generally agree:

**Better typing**: there are a lot of primitive types like string and number that can be better represented as either object types or string literals. Some examples of where this can be improved: Key signature (just [opened a PR](https://github.com/vexflow/vexflow/pull/199) for this), time signature, and position.

**Model more closely to MusicXML**: This is a bit of a theoretical standpoint, but since MusicXML is an official spec for sharing staff music, then I think it makes sense for Vexflow to better model this spec. An example of this (and could be improved regardless) is in the lack of a first-class Measure in Vexflow. I don't necessarily think Vexflow should convert directly from MusicXML to a rendered staff, nor do I think there is an ideal 1:1 mapping, but the closer they are modeled, I believe it will just lend itself to a better experience and an easier adoption of the library in general. This also aligns with the *better typing* comment where MusicXML already has well-defined types Vexflow can adopt (eg. [time element for TimeSignature](https://www.w3.org/2021/06/musicxml40/musicxml-reference/elements/time/).

**Responsive**: It would be great if Vexflow could have a default for width/height based on the size of the Element. I did this manually and it took me a lot longer than I expected because I had to infer some Vexflow math to get things displaying as I expected, which makes it really feel like this is Vexflow's responsibility.

**Comment by rvilarl** [comment]
@judocode I also like the idea of getting closer to XML keeping an eye on backwards compability. Help is always welcomed!

**Comment by mscuthbert** [comment]
If we're moving forward, you might also look at MNX as a guide (I am the editor of the MusicXML spec, but I'm also active in MNX development).  Downside: MNX is still in active development and we're changing things all the time; very little external support.  Upside: MNX is specifically designed to be a format that is easy for modern systems to implement and adapt to internal data structures.  It is designed to remedy exactly the things in MusicXML that are hard to parse or determine what the result is (no `<forward>` and `<backward>` tags).  Plus it's JSON which is much nicer for Vexflow.

**Comment by mscuthbert** [comment]
But in general, I don't think that this is a good next direction for Vexflow (maybe a v6 or v7 direction).  There are numerous systems that exist for getting data into Vexflow from formats such as MusicXML (my own music21j is one of them).  The biggest flaws in Vexflow that remain and can't be solved by external libraries are what does Vexflow do once it has all the data in its system.  How does it avoid layout collisions, compute beams, avoid lyric collisions, etc.  These are the directions I would prefer the community to focus on because they're the only ones that only Vexflow can solve.

**Comment by rvilarl** [comment]
@ronyeh @mscuthbert @AaronDavidNewman @jaredjj3 @alindsay55661  what about releasing a beta?

**Comment by jaredjj3** [comment]
My vote is to complete the following before a beta release:

- Implement all major features planned for Vexflow 5.
- Document all new features, especially font loading.
- Document breaking changes.
- Document the success criteria for promoting from beta to final.
- Announce beta release to [vexflow@googlegroups.com](mailto:vexflow@googlegroups.com).
- Redirect https://vexflow.com/tests to https://vexflow.github.io/vexflow-examples/tests/flow.html.

These may already be done or we may decide its unnecessary.

**Comment by mscuthbert** [comment]
I am -1 on having to complete a large set of tasks before releasing a beta at this point. In the past few years Vexflow has shifted to a cycle where the majority of dev time the main/HEAD branch is incompatible with the current release, so bug fixes and feature improvements can only be contributed by users working on the major refactor/infrastructure tasks. Breaking changes are needed from time to time, but I think it's a mistake to do more than the bare minimum between releases. Otherwise the remainder of the community--a large portion of which doesn't care about font changes or SMuFL or even Typescript--is shut out of contributing for large periods. 

This has been the case for me for the past several years. My projects now are filled with monkey patches to Vexflow 4 fixing bugs and layout that I will contribute to VF5 when it's released but in the midst of major refactors doesn't make sense to have two quite different solutions out there, neither one of which looks like it will make it to the public anytime soon.

**Comment by ronyeh** [comment]
I agree with more documenting and testing. 

I don't think there are any features that need implementing. We can bump features to 5.1.

I'm out until June 25. At that time I can help with beta and prod releases. In the meantime I'm sure you guys can manage additional prereleases if needed. 

Our next PRs should be for test pages and documentation. We can also use the wiki (and copy relevant docs from the v4 wiki as needed).

**Comment by ronyeh** [comment]
Did everything else seem to work? Did it just fail at the last part where
you publish to npm?

I think Mohit needs to add you as a maintainer on npm vexflow:

https://www.npmjs.com/package/vexflow

**Comment by rvilarl** [comment]
@0xfe could you add me?

**Comment by 0xfe** [comment]
Yes, I thought I already did it. Will do it now. :-)

**Comment by 0xfe** [comment]
Sent invitation.

**Comment by rvilarl** [comment]
@0xfe please send again and unexpected error occurred.

**Comment by 0xfe** [comment]
Looks like it worked, you show up as a maintainer on the repo.

**Comment by rvilarl** [comment]
@0xfe please send again and unexpected error occurred.

**Comment by rvilarl** [comment]
It did work, I published an alpha 4. Thanks.

@ronyeh the tag failed. My fault I did it from my fork. :(

**Comment by ronyeh** [comment]
Closing this discussion. 

Please continue the discussion at #224. There, we will make concrete steps toward releasing 5.0.
